File: tests/performance-test/src/main/js/WebSocketClientEcho.js
/*jslint node: true */

/*
 * #%L
 * %%
 * Copyright (C) 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

const args = process.argv.slice(2);
const port = args[0];
const times = args[1];
let receivedMessages = 0;
let remainingMessagesToSend = times;
let startTimestamp;

if (!port || !times) {
    console.log("usage: node WebSocketClientEcho.js port times");
    return;
}

const WebSocket = require("ws");
const ws = new WebSocket("ws://localhost:" + port);

// sample request message
const toSend = {
    _typeName: "joynr.JoynrMessage",
    customHeaders: {},
    header: {
        expiryDate: "1468570049406",
        msgId: "bbdf49f0-f585-44e6-b152-4812d5106547",
        from: "e4025dd3-ee2e-466e-b6ca-ba6cc8d20787",
        to: "4300528c-23ba-488e-8567-4bf344e084e9"
    },
    payload:
        '{"_typeName":"joynr.Request","methodName":"updateRoute","paramDatatypes":["joynr.types.Localisation.GpsPosition[]"],"params":[[{"_typeName":"joynr.types.Localisation.GpsPosition","latitude":11.65,"longitude":49.0065}]],"requestReplyId":"947d9e50-572e-43e6-9aca-865505b20e24"}',
    type: "request"
};

const killServerMessage = "killServer";
let totalTime;

var send = function(data) {
    remainingMessagesToSend--;

    if (remainingMessagesToSend >= 0) {
        ws.send(data, error => {
            if (error === undefined) {
                send(data);
            }
        });
    }
};

ws.on("message", (data, flags) => {
    receivedMessages++;

    if (receivedMessages >= times) {
        const elapsedTimeMs = Date.now() - startTimestamp;

        console.log("Elapsed time : " + elapsedTimeMs + " ms");
        console.log("Throughput   : " + times / (elapsedTimeMs / 1000.0) + " Msgs/s");

        ws.send(killServerMessage, () => {
            ws.terminate();
        });
        return;
    }
});

ws.on("open", () => {
    console.log("sending times:" + times);
    startTimestamp = Date.now();
    send(JSON.stringify(toSend));
});

File: tests/performance-test/src/main/cpp/websocket-client-echo/WebSocketClientEcho.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include <boost/program_options.hpp>
#include <chrono>
#include <cstdlib>
#include <functional>
#include <future>
#include <iostream>
#include <thread>
#include <websocketpp/client.hpp>
#include <websocketpp/config/asio_client.hpp>

using Client = websocketpp::client<websocketpp::config::asio_client>;

class BenchmarkTest
{
public:
    BenchmarkTest(Client& client, int numberOfMessagesToSend)
            : client(client),
              numberOfMessagesToSend(numberOfMessagesToSend),
              numberOfReceivedMessages(0),
              payload(R"({)"
                      R"("_typeName":"joynr.types.TestTypes.TStructExtended",)"
                      R"("tDouble":0.123456789,)"
                      R"("tInt64":64,)"
                      R"("tString":"myTestString",)"
                      R"("tEnum":"TLITERALA",)"
                      R"("tInt32":32)"
                      R"(})")
    {
    }

    void onConnectionEstablished(websocketpp::connection_hdl& connection)
    {
        startedTimestamp = std::chrono::high_resolution_clock::now();
        sendMessages(connection);
    }

    void onMessageReceived()
    {
        numberOfReceivedMessages++;

        if (finished()) {
            finishedTimestamp = std::chrono::high_resolution_clock::now();
        }
    }

    bool finished() const
    {
        return numberOfReceivedMessages >= numberOfMessagesToSend;
    }

    std::chrono::microseconds getDurationUs() const
    {
        return std::chrono::duration_cast<std::chrono::microseconds>(finishedTimestamp -
                                                                     startedTimestamp);
    }

    double getNumMessagesPerSecond() const
    {
        auto durationUs = getDurationUs();

        if (durationUs == std::chrono::microseconds(0)) {
            return 0.0;
        }

        return static_cast<double>(numberOfMessagesToSend) /
               (static_cast<double>(durationUs.count()) / 10e6);
    }

private:
    void sendMessages(websocketpp::connection_hdl& connection)
    {
        websocketpp::lib::error_code errorCode;

        for (int i = 0; i < numberOfMessagesToSend; i++) {
            client.send(connection, payload, websocketpp::frame::opcode::text, errorCode);

            if (errorCode) {
                std::cout << "Failed to send message #" << i << ": " << errorCode.message()
                          << std::endl;
            }
        }
    }

private:
    Client& client;

    int numberOfMessagesToSend;
    int numberOfReceivedMessages;

    std::chrono::high_resolution_clock::time_point startedTimestamp;
    std::chrono::high_resolution_clock::time_point finishedTimestamp;

    std::string payload;
};

int main(int argc, char* argv[])
{
    namespace po = boost::program_options;

    int port = 0;
    std::string hostAddress;
    int numberOfMessages = 0;

    po::options_description desc("parameters");

    desc.add_options()("help", "show usage")(
            "port,p", po::value<int>(&port)->default_value(4220), "server port")(
            "hostaddress,h", po::value<std::string>(&hostAddress)->default_value("localhost"),
            "server address")("numberofmessages,n",
                              po::value<int>(&numberOfMessages)->default_value(10000),
                              "number of messages to transmit");

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    if (vm.count("help")) {
        std::cout << desc << std::endl;
        std::exit(EXIT_SUCCESS);
    }

    try {
        Client client;

        client.init_asio();
        client.clear_access_channels(websocketpp::log::alevel::all);
        client.clear_error_channels(websocketpp::log::alevel::all);

        const bool secureConnection = false;
        websocketpp::lib::error_code errorCode;
        websocketpp::uri hostUri(secureConnection, hostAddress, port, std::string(""));

        BenchmarkTest benchmark(client, numberOfMessages);

        client.set_open_handler([&benchmark](websocketpp::connection_hdl connection) {
            benchmark.onConnectionEstablished(connection);
        });

        client.set_message_handler(
                [&benchmark, &client](websocketpp::connection_hdl connection,
                                      websocketpp::config::asio_client::message_type::ptr message) {
                    std::ignore = message;

                    benchmark.onMessageReceived();

                    if (benchmark.finished()) {
                        websocketpp::lib::error_code errorCode;
                        client.close(connection, websocketpp::close::status::normal,
                                     std::string(""), errorCode);

                        if (errorCode) {
                            std::cout << "Failed to close connection";
                            std::exit(EXIT_FAILURE);
                        }
                    }
                });

        auto connection = client.get_connection(hostUri.str(), errorCode);

        if (errorCode) {
            std::cout << "Failed to create connection: " << errorCode.message() << std::endl;
            std::exit(EXIT_FAILURE1);
        }

        client.connect(connection);

        auto thread = std::thread(&Client::run, &client);
        thread.join();

        std::cout << "Duration: " << static_cast<double>(benchmark.getDurationUs().count()) / 10e6
                  << " sec" << std::endl;
        std::cout << "Msgs/s: " << benchmark.getNumMessagesPerSecond() << std::endl;
    } catch (websocketpp::exception const& e) {
        std::cout << "Websocket++ exception: " << e.what() << std::endl;
    } catch (const std::exception& e) {
        std::cout << e.what() << std::endl;
    }

    return EXIT_SUCCESS;
}
